### ISSUE TITLE
Application Gateway - Adding child resource configurations with dynamic blocks causes conflicts

### ISSUE THREAD

**Issue by lotaezhao** [hashicorp/terraform-provider-azurerm#13468]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version
Terraform v1.0.7
azurerm v2.77.0

### Affected Resource(s)
azurerm_application_gateway

### Terraform Configuration Files

```hcl
resource "azurerm_application_gateway" "appgw" {
  name                = "${var.name_format}-appgw"
  resource_group_name = var.rgname
  location            = var.location
  tags                = var.tags

  sku {
    name     = var.app_gateway_sku_name
    tier     = var.app_gateway_tier
    capacity = var.app_gateway_capacity
  }

  gateway_ip_configuration {
    name      = "${var.name_format}-appgw-ip"
    subnet_id = var.subnet_id
  }

  frontend_port {
    name = local.frontend_http_port_name
    port = var.frontend_http_port
  }

  frontend_port {
    name = local.frontend_https_port_name
    port = var.frontend_https_port
  }

  frontend_ip_configuration {
    name                 = local.frontend_ip_configuration_name
    public_ip_address_id = azurerm_public_ip.pip.id
  }

  dynamic "backend_address_pool" {
    for_each = var.sites

    content {
      name  = "${var.name_format}-${backend_address_pool.key}-appgw-beap"
      fqdns = [ var.backend_address_pool_fqdns[backend_address_pool.key] ]
    }
  }

  backend_http_settings {
    name                                = local.http_setting_name
    affinity_cookie_name                = "${var.name_format}-appgw-affinity"
    cookie_based_affinity               = var.cookie_based_affinity
    path                                = var.backend_http_path
    port                                = var.backend_http_port
    protocol                            = var.backend_http_protocol
    request_timeout                     = var.backend_request_timeout
    pick_host_name_from_backend_address = var.backend_pick_host_name
  }

  dynamic "http_listener" {
    for_each = var.sites

    content {
      name                           = "${var.name_format}-${http_listener.key}-appgw-lstn"
      host_name                      = http_listener.value
      frontend_ip_configuration_name = local.frontend_ip_configuration_name
      frontend_port_name             = local.frontend_https_port_name
      protocol                       = "Https"
      ssl_certificate_name           = var.site_cert_names[http_listener.key]
    }
  }

  identity {
    type = "UserAssigned"
    identity_ids = [var.managed_identity_id]
  }
  
  dynamic "ssl_certificate" {
    for_each = var.key_vault_certificates

    content {
      name                = ssl_certificate.value
      key_vault_secret_id = var.key_vault_secret_ids[ssl_certificate.key].secret_id
    }
  }

  dynamic "request_routing_rule" {
    for_each = var.sites

    content {
      name                       = "${var.name_format}-${request_routing_rule.key}-appgw-rqrt"
      http_listener_name         = "${var.name_format}-${request_routing_rule.key}-appgw-lstn"
      backend_address_pool_name  = "${var.name_format}-${request_routing_rule.key}-appgw-beap"
      rewrite_rule_set_name      = "${var.name_format}-${request_routing_rule.key}-appgw-rwrl-st"
      rule_type                  = var.request_routing_rule_type
      backend_http_settings_name = local.http_setting_name
    }
  }

  ssl_policy {
    policy_type = var.ssl_policy_type
    policy_name = var.ssl_policy_name
  }

  dynamic "rewrite_rule_set" {
    for_each = var.sites

    content {
      name = "${var.name_format}-${rewrite_rule_set.key}-appgw-rwrl-st"

      rewrite_rule {
        name          = "${var.name_format}-${rewrite_rule_set.key}-appgw-rwrl"
        rule_sequence = 100

        condition {
          variable    = "http_resp_Location"
          pattern     = "(http|https):\\/\\/.*azurewebsites\\.net(.*)$"
          ignore_case = true
        }

        response_header_configuration {
          header_name  = "Location"
          header_value = "https://${rewrite_rule_set.value}{http_resp_Location_2}"
        }
      }
    }
  }
}
```

### Debug Output

Here is example of what occurs in terraform plan when adding site05 when site01,site02,site03,site04 already exist.

```hcl
      ~ http_listener {
          ~ host_name                      = "site01.org.nz" -> "site05.org.nz"
            id                             = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/rg-wordpress/providers/Microsoft.Network/applicationGateways/wordpress-appgw/httpListeners/wordpress-site01-appgw-lstn"
          ~ name                           = "ae-dev-wordpress-site01-appgw-lstn" -> "ae-dev-wordpress-site05-appgw-lstn"
            # (8 unchanged attributes hidden)
        }
      ~ http_listener {
          ~ host_name                      = "site02.org.nz" -> "site01.org.nz"
            id                             = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/rg-wordpress/providers/Microsoft.Network/applicationGateways/wordpress-appgw/httpListeners/wordpress-site02-appgw-lstn"
          ~ name                           = "ae-dev-wordpress-site02-appgw-lstn" -> "ae-dev-wordpress-site01-appgw-lstn"
            # (9 unchanged attributes hidden)
        }
      ~ http_listener {
          ~ host_name                      = "site03.org.nz" -> "site02.org.nz"
            id                             = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/rg-wordpress/providers/Microsoft.Network/applicationGateways/wordpress-appgw/httpListeners/wordpress-site03-appgw-lstn"
          ~ name                           = "ae-dev-wordpress-site03-appgw-lstn" -> "ae-dev-wordpress-site02-appgw-lstn"
            # (9 unchanged attributes hidden)
        }
      ~ http_listener {
          ~ host_name                      = "site04.org.nz" -> "site03.org.nz"
            id                             = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/rg-wordpress/providers/Microsoft.Network/applicationGateways/wordpress-appgw/httpListeners/wordpress-site04-appgw-lstn"
          ~ name                           = "ae-dev-wordpress-site04-appgw-lstn" -> "ae-dev-wordpress-site03-appgw-lstn"
            # (8 unchanged attributes hidden)
        }
      + http_listener {
          + frontend_ip_configuration_name = "wordpress-appgw-feip"
          + frontend_port_name             = "wordpress-appgw-https-feport"
          + host_name                      = "site04.org.nz"
          + name                           = "wordpress-site04-appgw-lstn"
          + protocol                       = "Https"
          + ssl_certificate_name           = "wildcard-something-com"
        }
```

### Expected Behaviour

Existing child resource configurations that haven't changed should be ignored and only the new child resources should be added.

### Actual Behavior
Terraform tries to replace existing child resources (http_listeners etc.) and encounters conflicts due to duplicate child resources. The behavior seems to be that child resources are returned as a list and it is shifting them all by 1 index position. The issue occurs for all the child resources created with dynamic blocks (backend_address_pools, http_listeners, request_routing_rules, rewrite_rule_sets, ssl_certificates etc.).

This often results in a conflict when it tries to create a new child resource with the same name as one that already exists causing the deployment to fail, as below:

```hcl
Error: creating/updating Application Gateway: (Name "ae-dev-wordpress-appgw" / Resource Group "ae-dev-rg-wordpress"): network.ApplicationGatewaysClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code="InvalidRequestFormat" Message="Cannot parse the request." Details=[{"code":"DuplicateResourceName","message":"Resource /subscriptions//resourceGroups//providers/Microsoft.Network/applicationGateways/ has two child resources with the same name (ae-dev-wordpress-site03-appgw-lstn)."}]
```

My current workaround is to delete all the child resources and let it recreate them all from scratch, but this causes an outage for existing applications behind app gateway.

### Steps to Reproduce

1. Deploy an application gateway with dynamic blocks to create a multi-site configuration (You probably need at least two sites to begin with to experience the behavior).
2. Then add an additional site into the configuration and redeploy.

### Important Factoids

N/A

### References

N/A

* #0000

**Comment by lotaezhao** [comment]
A related issue is: https://github.com/hashicorp/terraform-provider-azurerm/issues/9615 is requesting for application gateway child resources to be able to be created and managed outside the outside the application gateway resource.  This is badly needed since it is becoming common to build shared application gateways that load balance across many separate applications for cost efficiency.

As this is not possible at present there is no choice but to define the child resources as part of the application gateway resource. However, as per this issue if you use dynamic blocks this doesn't work correctly.

**Comment by Zeralith** [comment]
I just experienced near this exact same issue today! I added 2 new sites to our configuration and the Plan/Apply decided to just randomly shuffle around all the existing backend_address_pools/probes/http_settings/http_listeners/request_routing_rules/rewrite_rule_sets/etc. and have to delete/replace/recreate them.

The only difference, from @lotaezhao 's issue, was my Apply was able to resolve everything with no conflicts/errors, but the shuffling around of children resources is absolutely terrible, to say the least. This is unnecessary and could potentially cause an outage of already existing sites.

I don't think my Terraform code is absolutely necessary to post here, as it's much like @lotaezhao 's code posted above, but our Application Gateway's Module uses the following Dynamic blocks:

- dynamic "backend_address_pool"
- dynamic "probe"
- dynamic "backend_http_settings" (for non-wildcard sites, using an "if" to filter in the for_each)
- dynamic "backend_http_settings" (for wildcard sites, using an "if" to filter in the for_each)
- dynamic "http_listener" (for non-wildcard sites, using an "if" to filter in the for_each)
- dynamic "http_listener" (for wildcard sites, using an "if" to filter in the for_each)
- dynamic "redirect_configuration"
- dynamic "rewrite_rule"
- dynamic "request_routing_rule" (for non-wildcard HTTP sites, using an "if" to filter in the for_each)
- dynamic "request_routing_rule" (for wildcard HTTP sites, using an "if" to filter in the for_each)
- dynamic "request_routing_rule" (for non-wildcard HTTPS sites, using an "if" to filter in the for_each)
- dynamic "request_routing_rule" (for wildcard HTTPS sites, using an "if" to filter in the for_each)

azurerm_application_gateway should understand that if a child resource already exists then do NOT try to delete/replace/recreate it when a new site is added and when new child resources are needed to be created only. Definitely seems like some bad listing/IDing/indexing stuff is going-on and being committed to the State.

To have a good testing platform for fixing this issue, I would say to deploy/Apply an Application Gateway with a multi-site configuration with at least 6 sites. Mix-and-match with non-wildcard sites and with wildcard sites. Use some HTTP (80) to HTTPS (443) Redirects for some sites so they have 2x HTTP settings and Listeners per site. Have some sites use "Basic" Listener Types and then have some use "Multi site" Listener Types. Have some use the "Host Name from the HTTP Settings" and then some not. Use a few different Backend Pools where some are shared between some of the Request Routing Rules/Listeners and some that are just single use.

After that first deploy/Apply then, as a second deploy, add 6 more new sites with all of the above mixed-and-matched again, making sure that only the 6 new sites and all of their backend_address_pools/probes/http_settings/http_listeners/request_routing_rules/rewrite_rule_sets/etc. are added and none of the already existing ones, from the previous 6 sites, are deleted/replaced/recreated.

Lastly, in addition, this PR also needs to get approved and into a release. Not having the "Priority" on Request Routing Rules can break things when mixing-and-matching wildcard and non-wildcard sites, and sometimes we have no choice but to delete children and re-create things, due to wildcards needing to have a creation timestamp made after the non-wildcard sites: https://github.com/hashicorp/terraform-provider-azurerm/pull/13498

Would be great if all this was fixed in the AzureRM Provider, as soon as possible! Thank you, for everyone's hard work on this stuff!

**Comment by lotaezhao** [comment]
@Zeralith Thanks for your comments. Glad I'm not the only one whose encountered this!

In my case all the sites are just 'Wordpress on Linux' marketplace container running on App Service. To replicate the issue I think it would only take 3 or 4 sites to be configured and then add another using dynamic blocks.

The minimum child resources that should be needed are:
- dynamic backend_address_pool
- dynamic http_listener
- dynamic request_routing_rule
- http_settings

The dynamic blocks above iterate over a map containing the site names and site urls. So to add a new site I add it into the map and the code will deploy everything dynamically.

```hcl
sites = {
    "site01" = "site01.org.nz"
    "site02" = "site02.org.nz"
    "site03" = "site03.org.nz"
    "site04" = "site04.org.nz"
}
```

You would just need to deploy a number App Services (they don't have to contain anything) and use code similar to mine to deploy an app gateway in front of them using a map as the iterator for the dynamic blocks. Once done add a new site into the map and then the issue should occur. If I get time I will create a repo with simplified code that replicates the issue.

**Comment by tombuildsstuff** [comment]
Duplicate of #6896 - this is something we're planning to fix in 3.0 but is a breaking change, which is why that has to wait for a major release. In the interim providing a consistent sort order for these items, whilst not ideal - should remove the diff's.

**Comment by lotaezhao** [comment]
@tombuildsstuff Thanks for the reply. Yes I had thought using a fixed order would work but don't know a way of doing that with dynamic blocks. Terraform seems to be choosing the order, do you have an example of how to override that?

**Comment by Zeralith** [comment]
@lotaezhao From what I think I understand what @tombuildsstuff is saying would be something like,

"Don't use a generic module for this. Throw the idea of "Don't Repeat Yourself" out the window, right now... and you know that 400 line module you wrote? Well, turn that into like 10,000 lines now, because instead of using objects/maps/sets, you will now have to basically write a line-by-line "script" of exactly what you want to happen, with all the blocks duplicated and clearly written-out, instead of having things looped through at runtime. Lastly, when you add or delete sites in the future, instead of adding/deleting like 10 lines of "configuration code" that feed-in to the module, now you have to add/delete like 500 lines of "step-by-step code" directly within the module itself (so it's not a generic module anymore) and it's going to be spread-out all over the place within all the children blocks of azurerm_application_gateway. Everything all the other generic modules in your codebase do with their objects/maps/sets today, just stop doing that."

@tombuildsstuff what can we do to get this azurerm_application_gateway resource fixed and in the 3.0 beta as soon as possible? How do we vote/petition for it to be the very next resource to be added? Looking at the duplicate issues on this, I feel like there are a ton of beta testers chomping at the bit ready to start testing this... and this is the PR that looks like it will fix the issue (or maybe there needs more review and it's only like 90% of the way there?):

https://github.com/hashicorp/terraform-provider-azurerm/pull/7021

Lastly, just as a note, here are the resources that are in the 3.0 beta so far:

https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/3.0-upgrade-guide

Again, thank you for everyone's hard work on this stuff!

**Comment by lotaezhao** [comment]
@Zeralith Thanks, a little dramatic but nicely summarized ;-)

Yeah if I have to use a fixed order I defeat the whole purpose of what I'm trying to do here, which is create a fully elastic system for deploying Wordpress sites. I'm using Azure Pipelines to enable people who don't know Terraform to be able to add and remove Wordpress sites by simply updating a pipeline variable and running the pipeline (no code changes required). So I'm reliant on the dynamic features working here. If I code a workaround it will be messy and then I'll have to change it back again when the issue is fixed (which will destroy all the resources). Static blocks won't allow to achieve what I need, which is no code changes in order to add/remove websites. I can probably hold off for a while since I've only deployed the Dev env and cope with the appgw child resources being shuffled around and recreated. But once I hit production it would need to be working, if the fix was in a beta release I'd definitely be keen to test it.
